src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/AnalyzerWorkQueue.cs

```diff
@@ -0,0 +1,92 @@
+    {
+        private readonly int _throttlingMs = 300;
+
+        private readonly ConcurrentDictionary<ProjectId, (DateTime modified, ProjectId projectId, CancellationTokenSource workDoneSource)> _workQueue =
+            new ConcurrentDictionary<ProjectId, (DateTime modified, ProjectId projectId, CancellationTokenSource workDoneSource)>();
+
+        private readonly ConcurrentDictionary<ProjectId, (DateTime modified, ProjectId projectId, CancellationTokenSource workDoneSource)> _currentWork = 
+            new ConcurrentDictionary<ProjectId, (DateTime modified, ProjectId projectId, CancellationTokenSource workDoneSource)>();
+        private readonly int _timeoutForPendingWorkMs;
+        private ILogger<AnalyzerWorkQueue> _logger;
+
+        public AnalyzerWorkQueue(ILoggerFactory loggerFactory, int throttleWorkMs = 300, int timeoutForPendingWorkMs = 60*1000)
+        {
+            _logger = loggerFactory.CreateLogger<AnalyzerWorkQueue>();
+            _throttlingMs = throttleWorkMs;
+            _timeoutForPendingWorkMs = timeoutForPendingWorkMs;
+        }
+
+        public void PutWork(ProjectId projectId)
+        {
+            _workQueue.AddOrUpdate(projectId,
+                (modified: DateTime.UtcNow, projectId: projectId, new CancellationTokenSource()),
+                (_, oldValue) => (modified: DateTime.UtcNow, projectId: projectId, workDoneSource: oldValue.workDoneSource));
+        }
+
+        public ImmutableArray<ProjectId> TakeWork()
+        {
+            lock (_workQueue)
+            {
+                var currentWork = _workQueue
+                    .Where(x => x.Value.modified.AddMilliseconds(_throttlingMs) <= DateTime.UtcNow)
+                    .OrderByDescending(x => x.Value.modified) // If you currently edit project X you want it will be highest priority and contains always latest possible analysis.
+                    .Take(1) // Limit mount of work executed by once. This is needed on large solution...
+                    .ToImmutableArray();
+
+                foreach (var work in currentWork)
+                {
+                    _workQueue.TryRemove(work.Key, out _);
+                    _currentWork.TryAdd(work.Key, work.Value);
+                }
+
+                return currentWork.Select(x => x.Key).ToImmutableArray();
+            }
+        }
+
+        public void AckWorkAsDone(ProjectId projectId)
+        {
+            if(_currentWork.TryGetValue(projectId, out var work))
+            {
+                work.workDoneSource.Cancel();
+                _currentWork.TryRemove(projectId, out _);
+            }
+        }
+
+        // Omnisharp V2 api expects that it can request current information of diagnostics any time,
+        // however analysis is worker based and is eventually ready. This method is used to make api look
+        // like it's syncronous even that actual analysis may take a while.
+        public async Task WaitForPendingWork(ImmutableArray<ProjectId> projectIds)
```

Comment:
This API is confusing to me. Presumably the caller already has the mapping of ProjectId -> Task where the work is happening (because this class doesn't track Tasks).  When someone calls this, they're expected to also call AckWorkAsDone in order to get the cancellationTokens in line 80 to get cancelled?

Why not have this class also keep track of the actual tasks where work is being done?

Reply:
Case for ack and two different queues (current work and queue) is case where work is undergoing, its result are waited and in same time user writes new text and same project is queued again. However current waiting call should pass after currently analyzed version of project is done but it should not wait following one that has gone after it.

Case example (s):
* Project A is added.
* Results for project A is waited.
* Work is taken for analysis.
* User writes new content to project A -> project is queued again.
* Work for first analysis of A is ready.
* Api should return and not wait for another one.

* Project A, B is added.
* Results for project A, B is waited.
* B is taken under analysis, A stays in queue.
* User writes to A and B.
* A is taken under analysis
* A and B is done.
* Api should return and not wait pending B project but left it to next iteration.

And so on with more and more classes and order of events.

I will write more tests that opens those scenarios.

"Why not have this class also keep track of the actual tasks where work is being done?"

It could be write that way, but i don't think it will make it any simpler at least. It could return work object and keep track of them but how that makes solution as hole simpler? I think one way is store project snapshots and later document snapshots, but at the end only thing that changes at that point is that ProjectId is replaced with snapshot. I prefer that update if applicaple when i try implement analyzers against documents since documents are easier to create for unit testing queue. But that really doesn't change logic that much.

In this class taskCancellation isn't really used in context of tasks but it's composable way to write logic to support different scenarios of waiting events with timeout. It can be written with reset event's but at least few tries with them wen't more complicated solutions by far to support scenarios described above.

Reply:
Could naming as `WaitWorkReadyEvent` be more describing name?